wal/src/message_queue_impl/region.rs

```diff
@@ -200,13 +252,260 @@ impl<M: MessageQueue> Region<M> {
         ));
         let log_cleaner = LogCleaner::new(region_id, message_queue.clone(), log_topic);
 
+        info!(
+            "Finish to open region in namespace, namespace:{}, region id:{}",
+            namespace, region_id
+        );
+
         Ok(Region {
             inner,
             snapshot_synchronizer,
             log_cleaner,
         })
     }
 
+    async fn recover_region_meta_from_meta(
+        namespace: &str,
+        region_id: RegionId,
+        builder: &mut RegionMetaBuilder,
+        message_queue: &M,
+        meta_topic: &str,
+        meta_encoding: &MetaEncoding,
+    ) -> Result<Offset> {
+        info!(
+            "Recover region meta from meta, namespace:{}, region id:{}",
+            namespace, region_id
+        );
+
+        // Fetch high watermark and check.
+        let high_watermark = message_queue
+            .fetch_offset(meta_topic, OffsetType::HighWaterMark)
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(OpenWithCause {
+                namespace,
+                region_id,
+                msg: "failed while recover from meta",
+            })?;
+
+        // Meta topic is empty, it just need to recover from log topic.
+        if high_watermark == 0 {
+            debug!("Meta topic is empty, it just need to recover from log topic, namespace:{}, region id:{}", namespace, region_id);
+            return Ok(0);
+        }
+
+        // Fetch snapshot from meta topic(just fetch the last snapshot).
+        let mut iter = message_queue
+            .consume(meta_topic, StartOffset::At(high_watermark - 1))
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(OpenWithCause {
+                namespace,
+                region_id,
+                msg: "failed while recover from meta",
+            })?;
+
+        let (latest_message_and_offset, returned_high_watermark) = iter
+            .next_message()
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(OpenWithCause {
+                namespace,
+                region_id,
+                msg: "failed while recover from meta",
+            })?;
+
+        // TODO: maybe should assert it?
```

Comment:
I guess ensure is enough, because it is better not to assume the remote service always right.